prov/verbs: Implement verbs on Windows

dshinaberry opened the pull request:
Implement the verbs provider on Windows using an Infiniband Verbs to
NetworkDirect translation layer.

On-behalf-of: @cmru-ps <dshinaberry@mru.medical.canon>
Signed-off-by: Derek Shinaberry <dshinaberry@mru.medical.canon>

j-xiong commented on include/windows/infiniband/ib.h:
```diff
@@ -0,0 +1,100 @@
+	return ((a->sib_addr64[0] | a->sib_addr64[1]) == 0);
+}
+
+static /*inline*/ int ib_addr_loopback(const struct ib_addr *a);
```
comment out the `static` as well now that the function becomes external.

j-xiong commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -754,6 +754,7 @@ static void verbs_devs_free(void)
 
 static void vrb_set_peer_mem_support(void)
 {
+#ifndef _WIN32
```
It would be better to use `#ifdef` to choose from two versions of the same function rather than commenting out the entire function body.

dshinaberry replied:
These two functions were almost literally the first issues I encountered and I never looked closely at their usage. Now that I look more closely, they are only called from fi_verbs_ini, a.k.a. VERBS_INI. Assuming you are alright with the conditional compilation around vrb_nd_startup and vrb_nd_shutdown, the simpler thing would probably be to just leave both functions fully intact and conditionally exclude the calls to them by surrounding them with a `#ifndef _WIN32` directive.

j-xiong replied:
Yes, that is better.

j-xiong commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -772,10 +773,12 @@ static void vrb_set_peer_mem_support(void)
 
 static void vrb_set_dmabuf_support(void)
 {
+#ifndef _WIN32
```
Same here.

j-xiong commented on prov/verbs/src/windows/verbs_nd_ibv.c:
```diff
@@ -0,0 +1,871 @@
+//
+// infiniband/verbs.h defines ibv_reg_mr as a macro and then uses inline functions to perform some
+// optimization that should result in ibv_reg_mr_iova2 being called on Windows.
+// However, in order to link correctly, there must be a definition of ibv_reg_mr that is not
+// expanded by the macro from verbs.h.
```
Could you explain a little bit more? Why the definition is needed if never being used? Is there any place the non-macro version of ibv_reg_mr is referenced?

dshinaberry replied:
The shorter answer is that there is a call to it, but it is in a code path that will never be executed on Windows because that code path requires __builtin_constant_p to return true. It is a gcc builtin function and I defined it as a macro that always returns 0 for Windows. So, the call is there and requires something to link to, but that code path is dead for all intents and purposes on Windows.

This is a really wonky part of verbs.h. As background, one of my goals was to minimize edits to the header files I pulled from Linux and make the needed adaptations in new code. I could probably have made more substantial edits to verbs.h to avoid this and maybe that's what would be preferable here.

Here's the bit from verbs.h relating to ibv_reg_mr.
```
/**
 * ibv_reg_mr_iova2 - Register memory region with a virtual offset address
 *
 * This version will be called if ibv_reg_mr or ibv_reg_mr_iova were called
 * with at least one potential access flag from the IBV_OPTIONAL_ACCESS_RANGE
 * flags range The optional access flags will be masked if running over kernel
 * that does not support passing them.
 */
struct ibv_mr *ibv_reg_mr_iova2(struct ibv_pd *pd, void *addr, size_t length,
				uint64_t iova, unsigned int access);

/**
 * ibv_reg_mr - Register a memory region
 */
struct ibv_mr *ibv_reg_mr(struct ibv_pd *pd, void *addr, size_t length,
			  int access);
/* use new ibv_reg_mr version only if access flags that require it are used */
__attribute__((__always_inline__)) static inline struct ibv_mr *
__ibv_reg_mr(struct ibv_pd *pd, void *addr, size_t length, unsigned int access,
	     int is_access_const)
{
	if (is_access_const && (access & IBV_ACCESS_OPTIONAL_RANGE) == 0)
		return ibv_reg_mr(pd, addr, length, access);
	else
		return ibv_reg_mr_iova2(pd, addr, length, (uintptr_t)addr,
					access);
}

#define ibv_reg_mr(pd, addr, length, access)                                   \
	__ibv_reg_mr(pd, addr, length, access,                                 \
		     __builtin_constant_p(				       \
			     ((access) & IBV_ACCESS_OPTIONAL_RANGE) == 0))
```
What's your opinion as to what would be best? Better comments in verbs_nd_ibv.c while leaving verbs.h intact or edits to verbs.h to just reduce the whole thing to be a direct call to an ibv_reg_mr implementation and not have to comment about the weirdness.

j-xiong replied:
How about having this as the body of `ibv_reg_mr()` instead:
```
	return ibv_reg_mr_iova2(pd, addr, length, (uintptr_t)addr, access);
```
You can then remove those comments.

dshinaberry commented:
I'm looking into the CI failure now.

j-xiong commented:
efa has its own version of verbs.h (from https://github.com/aws/efawin/tree/v1.0.0/interface/infiniband).

shefty commented:
@ofiwg/aws-efa-team -- The purpose of this PR is to allow the verbs provider to run on windows over network direct, but through verbs shim calls (not a full layering of libibverbs).  Based on @j-xiong's command, it sounds like EFA has a similar mechanism defined for windows?  Is this an area that we can/need to coordinate on how to map the libibverbs API to windows?

wzamazon commented:
@gurugubs implemented efawin, which is a subset of rdma-core API meant to support EFA on window.

I will let him answer question about coordinating on how to map the libibverbs API to windows?

wzamazon commented:
I wonder why `verbs.h` is in `src/include/windows`, not inside `prov/verbs/src/windows/`. Where the implementation resides?

j-xiong commented:
> I wonder why `verbs.h` is in `src/include/windows`, not inside `prov/verbs/src/windows/`. Where the implementation resides?

I think you are right. `prov/verbs/src/windows` is a better place for all these headers unless they can be shared by multiple providers.

dshinaberry commented:
To be honest, I wasn't even aware of the efa usage of anything related to verbs on windows. I had seen the include/windows directory tree for the entire libfabric library and simply assumed that was the best place to put the header files.

The prov/verbs/src/windows directory only came about after a discussion with @shefty about where to put the windows specific files related to the verbs provider. I wasn't adding them to the makefile used on linux at all, so it was more a matter of where to put the files in the directory hierarchy to make clear they were windows specific.

I could certainly relocate the new header files to somewhere under prov/verbs/windows if that's what the maintainers would like. I also had to make a few changes to files that were already under include/windows. Not sure whether we would keep those there or clone those files to be under prov/verbs/src/windows as well.

As to whether we can or should harmonize between verbs and efa in the area of how they abstract the verbs API for use on windows, I'll defer to the maintainers because that seems like a project strategy/roadmap sort of concern. I'm just a contributor and am happy to adapt my offering to suit the maintainers' desires for the project.

j-xiong commented:
Only the `infiniband` and `rdma` subdirectories need to be relocated since they are strictly for Verbs only. Other files under `include/windows` should stay.

dshinaberry commented:
I didn't stop and think through the details of just moving the location of the header files. That allows everything to compile, but then the link fails because the ibv_* and rdma_* functions are multiply defined - once by the new verbs code and once by the efa code.

It looks to me like the efa code is only used in the two specific configurations `Debug-Efa-v142` and `Release-Efa-v142`. Is this provider completely independent of the verbs provider? In other words, could we omit the verbs provider from these two configurations? That could be quick solution to the issue at hand.

The not so quick solution would be to figure out how to combine them harmoniously or make use of some sort of trampoline indirection sort of like all the 'ops' structures used at the top level of libfabric.

I know that there are providers that layer over other providers and some sort of hook provider mechanism, but, to be honest, I don't understand how those work. But could we make use of those somehow as a template for something that could allow verbs and efa to utilize the verbs API with different backing ND and EFA respectively?

wzamazon commented:
> Is this provider completely independent of the verbs provider?

Yes. EFA is independent of verbs provider.

> In other words, could we omit the verbs provider from these two configurations? 

Yes, and we should do that.

@gurugubs Can you help with the change?

shefty commented:
Before forcing providers to be exclusive with each other, I'd like to understand what the 2 providers are doing and focus on a shared solution. We're trying to provide some linux specific APIs on windows for portability.  I don't see that this is really any different than what we have for any other portability files in src/windows or include/windows.  I have not yet looked at this PR in details.  What functions need to be added to windows?  How are those functions implemented?

gurusrik commented:
The API defined by verbs.h is a mirror of the API surface from rdmacore on Linux(OpenIB verbs interface).  rdmacore on linux offers ability for multiple providers to be used with it similar to how libfabric supports various providers to serve the same libfabric API.  rdmacore is not available on Windows and porting the entire framework to support providers on windows the way it does on Linux would be a significant effort.  For EFA we ported the efa provider of rdmacore to Windows and hooked the handlers directly to the subset of verbs api that the libfabric efa provider uses.  @dshinaberry appears to have done exactly the same thing for ND resulting in a conflict.  The way to resolve this for all future verbs API users would be to introduce a layer that routes the API to different handlers based on a build/run time configuration variable.  Given the effort and time involved in making such an implementation, it is probably still better to exclude the verbs provider from EFA configurations to facilitate forward progress.

shefty commented:
I know about rdma-core and the libibverbs architecture.  How does the efa verbs provider on windows communicate with the kernel driver?

Most of the function definitions in verbs.h are defined to go through provider specific handlers.  Why can't we just use that?

dshinaberry commented:
I was/am not familiar at all with the rdma-core and libibverbs architecture. I essentially gutted all inline implementations from infiniband/*.h and rdma/*.h header files and crafted implementations that invoke ND functions that now reside in prov/verbs/src/windows/verbs_nd_{ibv,rdma}.c.

From: Sean Hefty ***@***.***>
Sent: Monday, March 28, 2022 12:38
To: ofiwg/libfabric ***@***.***>
Cc: Shinaberry, Derek ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [ofiwg/libfabric] prov/verbs: Implement verbs on Windows (PR #7596)


I know about rdma-core and the libibverbs architecture. How does the efa verbs provider on windows communicate with the kernel driver?

Most of the function definitions in verbs.h are defined to go through provider specific handlers. Why can't we just use that?

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/ofiwg/libfabric/pull/7596*issuecomment-1080951542__;Iw!!Ai2CFrZnFhI!67_ANHTs3sZniHkcYP-iLXFJhSTs_e4dqc53f0-kqOzmnZ5eOb4Ndx_cfCfgPELr9_KxQghTuKyKPvZDarXEP7vb9EPJt0MzOg$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AU5GZ4UUJPONULZGKO2PFYDVCHU7BANCNFSM5RRSBB4Q__;!!Ai2CFrZnFhI!67_ANHTs3sZniHkcYP-iLXFJhSTs_e4dqc53f0-kqOzmnZ5eOb4Ndx_cfCfgPELr9_KxQghTuKyKPvZDarXEP7vb9EMY2RPMSw$>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

gurusrik commented:
EFA provider uses the handler functions defined in efawin library that we copy over for Windows builds similar to how we copy NetworkDirect related artifacts.  Since there is no rdmacore library on Windows to handle routing to provider specific handlers, we added shim functions that implement verbs API that route to the efa handlers.  To use this, we need the shim functions to be more sophisticated and probably handle a lot more cases than were necessary for efa provider.

wzamazon commented:
@shefty My understanding is that the header files from rdma-core repo do not compile on windows. To make them compile on windows requires significant amount of work. (so @gurusrik and @dshinaberry have each extracted part of the headers they needed to support the provider).

@dshinaberry Reading the `libfabric.vcxproj` it seems what you worried (duplicated symbol) will not happen. Because the verbs source code is not included in EFA windows build, vice versa.

dshinaberry commented:
So, it's been a couple of days and I'm not clear on what is wanted in order to move forward with this PR and get it approved.

It seems like @wzamazon and @gurusrik were fine with the idea of of removing the verbs provider from the efa specific configurations in order to move forward. Presumably we'd then create an issue to investigate/consider what changes would be needed to allow the two providers to coexist in the same build.

It seems like @shefty wasn't prepared to go down this path just yet.

The most recent commit I pushed disables the verbs provider in the efa specific configurations. So, it is in line with what @wzamazon and @gurusrik thought was acceptable.

@shefty, do you just need more time to consider the pros and cons? Or would you prefer me to make some changes to the current PR? Or would you like me to do some investigating and report back?

shefty commented:
I just haven't had time to analyze this PR yet.  So I would hold off making changes.  I also don't necessarily see that the solutions need to be independent.  We could pick one as a short term fix while working toward a better alternative.

shefty commented on prov/verbs/src/windows/fi_verbs_nd.h:
```diff
@@ -0,0 +1,259 @@
+	vrb_nd_event_base base;
+	IND2Connector *connector;
+	struct rdma_cm_id *listen_id;
+} vrb_nd_cm_listen_event;
```
Please no typedefs for structures or enums.  Just use struct blah where this is needed.  Typedefs require someone reading the code to determine what the type actually is and translate.  In general typedefs are only useful where the underlying type is actually hidden from the user because the implementation may change based on the platform.

shefty commented on prov/verbs/src/windows/fi_verbs_nd.h:
```diff
@@ -0,0 +1,259 @@
+	COMPLETE,
+	DISCONNECT,
+	DISCONNECTED
+} vrb_nd_cm_state;
```
No typedef and please add a prefix to the values.  There should be some IDLE, UNCONNECTED, or UNKNOWN state as the first value, since a newly allocated structure that uses this enum will not start as connected.

shefty commented on prov/verbs/src/windows/fi_verbs_nd.h:
```diff
@@ -0,0 +1,259 @@
+		uint8_t responder_resources;
+		uint8_t initiator_depth;
+	} param;
+} vrb_nd_cm_connect_event;
```
no typedef

shefty commented on prov/verbs/src/windows/fi_verbs_nd.h:
```diff
@@ -0,0 +1,259 @@
+	vrb_nd_event_base base;
+	struct ibv_comp_channel_vrb_nd *ch_nd;
+	struct ibv_cq *cq;
+} vrb_nd_cq_notification;
```
no typedef

shefty commented on prov/verbs/src/windows/fi_verbs_nd.h:
```diff
@@ -0,0 +1,259 @@
+inline void trace(const char *function)
+{
+	VRB_INFO(FI_LOG_FABRIC, "%s\n", function);
+}
```
Use VRB_INFO directly, or VRB_TRACE (if that exists) since this is tracing a function call.  The logging macros automatically log the function name, so calling the FI_LOG macro with an empty string is sufficient.

shefty commented on prov/verbs/src/windows/fi_verbs_nd.h:
```diff
@@ -0,0 +1,259 @@
+		}
+	}
+	return NULL;
+}
```
Most functions declared in a header file should be static inline

shefty commented on prov/verbs/src/windows/fi_verbs_nd.h:
```diff
@@ -0,0 +1,259 @@
+	return NULL;
+}
+
+inline struct ibv_context *get_context_by_device(struct ibv_device *device)
```
static inline

shefty commented on prov/verbs/src/windows/fi_verbs_nd.h:
```diff
@@ -0,0 +1,259 @@
+inline void unimplemented(const char *function)
+{
+	VRB_WARN(FI_LOG_FABRIC, "%s\n", function);
+}
```
use VRB_WARN directly -- not sure what this function is intended to do though.

shefty commented on prov/verbs/src/windows/infiniband/ib.h:
```diff
@@ -0,0 +1,100 @@
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
```
Should document the source of this file for reference.

shefty commented on prov/verbs/src/windows/infiniband/ib.h:
```diff
@@ -0,0 +1,100 @@
+	return ((a->sib_addr32[0] | a->sib_addr32[1] |
+		 a->sib_addr32[2] | (a->sib_addr32[3] ^ htobe32(1))) == 0);
+}
+*/
```
We can just delete dead code

shefty commented on prov/verbs/src/windows/infiniband/ib_user_ioctl_verbs.h:
```diff
@@ -0,0 +1,212 @@
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
```
Document source of the file for reference, but I'm curious.  Are all of the definitions in this file needed?

dshinaberry replied:
Probably not. If you'd like I can spend some more time pruning out the unused declarations/definitions from the headers I pulled from rdma-core.

shefty replied:
We could also treat this as post merge cleanup.

shefty reviewed (commented):
I've reviewed 11 of 18 of the files.  This is just a lot of code to go through.  Will try to get through the rest of the files in the next couple of days.  See comments on the files reviewed so far.

shefty commented on prov/verbs/src/windows/verbs_nd_rdma.c:
```diff
@@ -0,0 +1,710 @@
+			adapter = vrb_nd_infra.adapters.adapter_list[i];
+			id->route.addr.src_addr = *addr;
+			break;
+		} else if (!strncmp(name, "127.0.0.1", strlen(name)) &&
```
nit: else not needed since if statement breaks.

shefty commented on prov/verbs/src/windows/verbs_nd_rdma.c:
```diff
@@ -0,0 +1,710 @@
+		return -1;
+	}
+
+	if (adapter) {
```
Exiting the function early on !adapter would outdent most of the code.

shefty commented on prov/verbs/src/windows/verbs_nd_rdma.c:
```diff
@@ -0,0 +1,710 @@
+		}
+	}
+
+	id->route.addr.dst_addr = *dst_addr;
```
sockaddr isn't a real address.  We should check the sa_family and copy the address based on the size sockaddr_in or sockaddr_in6 or sockaddr_ib)

shefty reviewed (changes requested):
Finished review of all files.  Overall, this is looking really good.  See comments for changes and request for coding style changes.

shefty commented on include/ofi_prov.h:
```diff
@@ -59,6 +59,7 @@ GNI_INI ;
 #  define GNI_INIT NULL
 #endif
 
+#if !defined(_WIN32) || !defined(HAVE_EFA)
```
The comment and restructuring definitely helps here.  Thanks!

dshinaberry commented:
I have a couple of process related questions regarding your comments.

1. I pulled some of the initialization code directly from the netdir provider. If a comment applies to that code, do you want a modification in the original netdir code as well?
2. @j-xiong had made a remark previously about squashing commits that update a PR. I can understand only wanting to merge a single commit. However, having distinct commits when updating a review can make it a lot easier to see just the updates in relation to review comments. I just wanted to get your take on how you would like me to update the PR when I've addressed all the concerns.

shefty commented:
The code in the netdir is in poor shape.  I would not bother about updating it.  Once this is merged, I want to see if we can drop the netdir provider completely.  But I need to check with our MPI team before doing that and ensure there aren't other users dependent on the netdir provider protocol.

As for commits that update a PR, the changes should be made to the commits directly (so squashed).  We need that for merging and git-bisect purposes.  (We don't want to merge a broken patch.)  If you want to create new commits with the changes for review purposes, that's fine.  But we'll ask for them to be squashed prior to merging.  Just add a comment to the commit or github with that intent.

dshinaberry commented on prov/verbs/src/windows/fi_verbs_nd.h:
```diff
@@ -0,0 +1,259 @@
+	const char *name;
+	struct ibv_device device;
+	struct ibv_context context;
+};
```
If, by anywhere, you mean in standard Windows header files, then no. These structures and most of the code in vrb_nd_init.c was taken from the netdir provider source code.

dshinaberry commented on prov/verbs/src/windows/infiniband/verbs.h:
```diff
@@ -0,0 +1,3565 @@
+#  undef __attribute_const
+
+
+#endif /* INFINIBAND_VERBS_H */
```
See my comment above about whether you'd like me to invest some time in pruning the imported header files.

As it stands, I only provided implementations for methods that the verbs provider actually uses and even a handful of those simply return ENOSYS at the moment.

If a declared function that isn't implemented on Windows were to be added to the verbs code, it would result in a link error due to an undefined symbol. Not quite a compilation failure, but close.

dshinaberry commented on prov/verbs/src/windows/verbs_nd_ibv.c:
```diff
@@ -0,0 +1,857 @@
+#define VRB_ND_MAX_CQ_CNT (1 << 24)
+#define VRB_ND_MAX_QP_CNT (1 << 18)
+#define VRB_ND_MAX_SRQ_CNT (1 << 23)
+#define VRB_ND_MAX_MR_CNT (1 << 24)
```
Fair point about the magic numbers being specific to my test hardware. Short of collecting numbers from a bunch of different hardware and using the lowest common denominator, I'm not sure what else we could do. And I don't have access to a variety of hardware.

I can be more specific about the hardware I collected the numbers from if you think that adds any value.

ND does support SRQ, so I think a non-zero value makes sense.

shefty replied:
If ND supports SRQ, then we probably need to figure out if the HW supports SRQ somehow.  Or does ND require SRQ support?  Because it's optional for RDMA devices.

shefty replied:
Btw, I don't think anything in libfabric or any users of libfabric will end up caring about most of these values.  So anything reasonable should be okay.  The max_srq is an exception that I'm aware of.

dshinaberry commented on prov/verbs/src/windows/verbs_nd_ibv.c:
```diff
@@ -0,0 +1,857 @@
+	}
+
+	struct ibv_comp_channel *channel = (struct ibv_comp_channel *)calloc(
+		1, sizeof(struct ibv_comp_channel_vrb_nd));
```
I'll switch around the places where I allocate memory. That's easy enough. Do you have a preference for how to obtain a pointer to the outer structure when a pointer to the inner structure is passed in as a parameter?

This is very much like a derived class in C++ where a pointer to base is returned. Since there is no dynamic_cast in C to obtain the pointer to derived again, it has to be just a cast. It seemed like I saw container_of being used frequently in this way, but I can just use straight casts if you think that would be clearer.

shefty replied:
Using container_of to get an outer structure from an inner one is what we want.  The problem here is that the code is allocating the size of the *outer* structure, but assigning the memory to a pointer that's an *inner* structure.  It then uses container_of to get the outer structure.  This flow only works if the inner structure is first member of the outer.  I'm only referring to this flow; there's no actual issue with using container_of.  That's ideal.

dshinaberry commented on prov/verbs/src/windows/verbs_nd_ibv.c:
```diff
@@ -0,0 +1,857 @@
+	while (channel_nd->num_pending) {
+		int ret = dlistfd_wait_avail(&channel_nd->q, -1);
+		if (ret == 1) {
+			ofi_mutex_lock(&channel_nd->q_lock);
```
I'll admit the num_pending, q_lock, and q members may be a bit confusing.

When ibv_req_notify_cq is called, num_pending is incremented because there will eventually be an entry added to q, but not until the asynchronous notify completes either successfully or in error, e.g. cancelled during teardown.

When ibv_destroy_comp_channel is called, we need to wait until all asynchronous requests are completed and add an entry to q. Thus the check for num_pending > 0 and not q empty.

I think the lock for q is needed, because ibv_destroy_comp_channel will be competing with nd_cq_notify_event/nd_cq_notify_error for access to the queue. The former trying to drain it and the latter trying to add entries.

So, I think both num_pending and q are needed and the logic is correct. I am going to change num_pending into an atomic though as there could be contention for it.

This was a tricky spot though, so please check my math. ;-)

shefty replied:
If another thread is trying to acquire the mutex, then it could try to do so after mutex_destroy (followed by free) has been called below, resulting in an invalid memory access.  The protection here isn't clear to me.

What is num_pending counting?  What increments the value and what decrements the value do not seem to be related.

dshinaberry replied:
Let me take another crack at explaining this to see if I can make it make sense. The context for all this is a completion channel.
```
ibv_create_comp_channel
    initialize queue
    initialize mutex
    initialize num_pending to 0

ibv_req_notify_cq
    starts an asynchronous call that will lead to a callback to either
        nd_cq_notify_event or
        nd_cq_notify_error
    increments num_pending indicating there is a new outstanding call

nd_cq_notify_event/nd_cq_notify_error
  lock mutex
  insert completion event into the queue
  unlock mutex
```
In the normal course of events `ibv_get_cq_event` will be either polled or called after an epoll wait on the channel's fd returns.
```
ibv_get_cq_event
  lock mutex
  remove completion event from the queue
  decrement num_pending
  unlock mutex
```
Now lets look at what occurs during a shutdown scenario. In this case there will be a situation where the queue is empty, but `num_pending` is 1 because there is an outstanding notification request that hasn't completed.

The client code will call `ibv_destroy_cq`, which will trigger any outstanding requests to complete with a cancelled status, and then `ibv_destroy_comp_channel`, which wants to free all the associated resources.

So this sets up a race between `nd_cq_notify_error` and `ibv_destroy_comp_channel` where we want to force a wait to allow the outstanding requests to complete and insert an entry into the queue.

Looping until `num_pending` is 0 does this.
```
ibv_destroy_comp_channel
    while num_pending > 0
        wait for non-empty queue
        lock mutex
        remove completion event form the queue
        decrement num_pending
        unlock mutex
    end loop
```

So, there can only be contention for the mutex and queue if `num_pending > 0`. Since we loop until `num_pendng == 0` we will never destroy the mutex and queue and free the memory if there might be contention.

So, the logic still seems sound to me, but there is always a chance I'm wrong. Clearly it's complicated given the back and forth. So, pick it apart and point out any shortcomings. Also, if you think there is a clearer way to achieve the same result please share.

shefty replied:
One issue is that num_pending isn't directly related to what it is intending to track, which is the number of outstanding overlapped notify operations.  Once nd_cq_notify_event/nd_cq_notify_error have been called, the overlapped operation is done.  That's where num_pending should be decremented.  Plus the name num_pending isn't descriptive enough for someone to figure out what is actually pending.

Another issue is that the behavior of ibv_req_notify_cq is to 'arm' the cq to generate an event on the next completion.  While the cq is armed, additional calls to ibv_req_notify_cq are essentially no-ops.  There should only be 1 event generated for a CQ until it is re-armed after an event has been generated, not 1 event per ibv_req_notify_cq call.

Similarly, we should only have at most 1 outstanding notify request.  So this could be simplified by adding the overlapped structure directly to the cq with a flag to indicate if it's in use.  Destroy could check this, and if it's in use, cancel the overlapped, wait for it to complete, then cleanup the event list.  It's been too long since I've looked at ND and overlapped operations that I don't remember how to cancel them, but I assume there's a way.

dshinaberry replied:
The ND documentation for CompletionQueue::Notify says: 
> Multiple requests for notification can be outstanding for the same completion queue. All such requests will be completed by the next completion event on the specified completion queue. This is similar to a notification event that releases all waiting requests, rather than a synchronization event that releases, at most, one. This allows multiple threads to process completions in parallel.

I think this is why I implemented it as I did. That being said, I suppose I could boil it down to match the behavior you describe for the Linux ibverbs implementation since that is what we're trying to emulate.

However, even after that, I don't think it can be quite as simple as you describe. There can be a many to one relationship between CQs and completion channels. And the data structures point towards a CQ knowing its channel, but a channel not knowing its CQs.

The documentation for ibv_destroy_comp_channel mentions that it will fail if any CQs are still associated with it. I must not have taken notice of that, because I sure didn't attempt to implement it that way. I also think I recall seeing a refcnt member in ibv_comp_channel. So perhaps the best way is not to count outstanding notify requests, but use the refcnt member to return an error if it is non-zero and if it's zero, then proceed to drain the queue of any events and release all the resources.

shefty replied:
My comment above is relative to the cq, not the comp_channel.  My reference to 'destroy' is to ibv_destroy_cq.  For the comp_channel, yes, I believe you can use a simple refcnt for the number of associated cq's and fail if not zero.

What I was describing above is adding an overlapped structure to the cq, along with some flag.  When ibv_req_notify_cq () is called, mark the overlapped as in-use.  When nd_cq_notify_event/nd_cq_notify_error is called, mark the overlapped as available.  In ibv_destroy_cq, see if the overlapped is in-use, and if it is, cancel it and wait for the cancel to finish.

dshinaberry commented on prov/verbs/src/windows/verbs_nd_ibv.c:
```diff
@@ -0,0 +1,857 @@
+	int ret = CloseHandle(cq_nd->ov_file);
+	assert(ret);
+	// Can't assert that the reference count is zero because releasing the
+	// associated QueuePair doesn't always release its CompletionQueues.
```
What you can't read my mind? ;-)

I added this comment when I was working on making sure that all acquired resources were released. I think it's a bug in the Mellanox ND driver for my hardware.

I could update the comment to something like...
```
    // There appears to be a bug related to the reference counting for
    // ComplettionQueue objects in the Mellanox ND driver. CompletionQueue
    // reference counts are incremented when they are used in the creation
    // of a QueuePair, but are not always decremented when the QueuePair is
    // Released.  Ideally, we'd assert that that the reference count is zero
    // here, but we cannot because of that bug.
```
...or just drop it altogether.

shefty replied:
If there's a reproducible bug in the lower SW that we need to work around, I'd go with the comment you proposed.  Otherwise there's no way for a reader to figure out why the code is structured the way it is.

dshinaberry commented:
Alright, I think I have all the requested changes in place and squashed it all down to a single commit. Let's see what I might have missed...

shefty commented on prov/verbs/src/windows/verbs_nd_init.c:
```diff
@@ -0,0 +1,751 @@
+int ofi_nd_is_directory(const wchar_t *path);
+
+static size_t unique(void *base, size_t num, size_t width,
+		     int (*cmp)(const void *, const void *))
```
The function parameters and name don't lend themselves to figuring out what this call does.

dshinaberry replied:
This came straight from the netdir provider. Would a comment above the function like the following be sufficient?
```
/*
 * Assuming a sorted array,' base[num]', with elements of size 'width', this
 * function will compare adjacent elements using 'cmp()' and "compact" the
 * array to just its 'n' unique elements.
 */
```

shefty replied:
I would restructure to something like this then:
```
/* Array must be sorted */
static void ofi_nd_remove_dups(void *array, size_t *entries or count, size_t elem_size, ...)
```
This makes it a little clearer as to the intent.  There's still some confusion here.  If we're only removing duplicates and the array is sorted, why do we need the cmp function passed in?  memcmp would work

shefty commented on prov/verbs/src/windows/verbs_nd_init.c:
```diff
@@ -0,0 +1,751 @@
+		OFI_NDERR(ND_HOST_UNREACHABLE, err_str);
+		OFI_NDERR(ND_CONNECTION_ABORTED, err_str);
+		OFI_NDERR(ND_DEVICE_REMOVED, err_str);
+		OFI_NDERR(ND_DISCONNECTED, err_str);
```
Please outdent the OFI_NDERR macros 1 level

shefty commented on prov/verbs/src/windows/verbs_nd_init.c:
```diff
@@ -0,0 +1,751 @@
+	if (err != WSAEFAULT || !len)
+		return NULL;
+
+	prov = (wchar_t *)malloc(len * sizeof(*prov));
```
Like calloc, no need to cast malloc() results.

shefty commented on prov/verbs/src/windows/verbs_nd_init.c:
```diff
@@ -0,0 +1,751 @@
+	// Allocate memory for adapter_list, device_list, and context_list all at once.
+	nd_infra.adapters.adapter_list =
+		(struct nd_adapter **)calloc((nd_infra.adapters.count + 1) * 3,
+					     sizeof(nd_infra.adapters.adapter));
```
We're mixing 3 different structures here and using a sizeof() value that isn't obviously related to any of them.  I think this should be something like:
```
calloc(nd_infra.adapters.count + 1,
    sizeof(*nd_infra.adapters.adapter_list) +
    sizeof(*nd_infra.adapters.device_list) +
    sizeof(*nd_info.adapters.context_list));
```

shefty commented on prov/verbs/src/windows/verbs_nd_init.c:
```diff
@@ -0,0 +1,751 @@
+		adapter->context.num_comp_vectors = 0;
+	}
+
+	return S_OK;
```
This function has a lot of memory allocations and early exit paths.  Please use a common exit path for errors, so that cleanup of all allocations is handled properly.

shefty commented on prov/verbs/src/windows/verbs_nd_init.c:
```diff
@@ -0,0 +1,751 @@
+			nd_create_factory(&proto[i]);
+	}
+
+	free(proto);
```
If you drop this line...

shefty commented on prov/verbs/src/windows/verbs_nd_init.c:
```diff
@@ -0,0 +1,751 @@
+	   create adapters & collect adapter's info */
+	hr = nd_create_adapter();
+
+	return hr;
```
...you an drop this line, and fall through below.

shefty commented on prov/verbs/src/windows/verbs_nd_init.c:
```diff
@@ -0,0 +1,751 @@
+	return hr;
+}
+
+static int nd_startup_done = 0;
```
bool?

shefty commented on prov/verbs/src/windows/verbs_nd_ov.c:
```diff
@@ -0,0 +1,399 @@
+	       hr);
+	entry_nd->event.param.conn.private_data_len = (uint8_t)len;
+	if (len) {
+		entry_nd->event.param.conn.private_data = malloc(len);
```
see prior comment

dshinaberry replied:
Hmm, not sure how I missed these. Sorry about that.

dshinaberry replied:
So, upon closer inspection, I see that there are a number of allocations in this file that don't check for success. There is an assert that it succeeded, so if asserts are enabled the program will terminate. However, if asserts are not enabled, the code will just continue merrily along and crash due to a null pointer deference.

Do you have an opinion as to whether this is alright or should there be code to check, log a warning message, and terminate when asserts are disabled?

Given that these are all asynchronous callback routines, I can't think of a graceful way to handle a memory allocation failure.

shefty replied:
Is there any way to allocate the memory prior to the callback routines being invoked, e.g. as part of the overlapped structure?  If the allocation fails in this specific case, can we set the data to null and len to 0 instead?  This is really unlikely, so starting with an assert and adding a TODO could work for now.  I know that static code analysis tools will flag this.

shefty commented on prov/verbs/src/windows/verbs_nd_ov.c:
```diff
@@ -0,0 +1,399 @@
+	       "IND2Listener::GetConnectionRequest: hr=0x%08lx; ov=%p\n", hr,
+	       &conn->base.ov);
+
+	free(base);
```
see prior comment

dshinaberry replied:
So this was the pattern that netdir was using, which I largely left intact, to deal with asynchronous overlapped requests and completions.

The requestor would allocate a buffer with an OVERLAPPED structure at the beginning, followed by functions to call for success/failure, and then any additional data that was needed by the completion routine to process things.

All completions were routed to a common callback function, nd_io_cb, then handled from there as needed.

Upon completion, the completion routine would free the buffer that had been used.

Would you be alright with just adding comments at the allocation and free locations about the usage pattern or would you prefer some sort of structural changes?

shefty replied:
I'd prefer some sort of structural change so that the lifetime of the objects is clear.  Let's assume whatever the current netdir provider is doing is likely wrong and unmaintainable.  :)

shefty commented on prov/verbs/src/windows/verbs_nd_ov.c:
```diff
@@ -0,0 +1,399 @@
+	dlistfd_insert_tail(&entry_nd->entry, &ch_nd->q);
+	VRB_DBG(FI_LOG_EQ, "EQ SET event:%p %p %d\n", entry_nd->event.id,
+		entry_nd->event.listen_id, entry_nd->event.event);
+	free(base);
```
see prior comment

shefty commented on prov/verbs/src/windows/verbs_nd_ov.c:
```diff
@@ -0,0 +1,399 @@
+		       hr, &disconnect->base.ov);
+	} else if (event->state == DISCONNECTED) {
+		entry_nd->event.event = RDMA_CM_EVENT_DISCONNECTED;
+	}
```
See prior comments for this function.

shefty reviewed (commented):
See comments based on this round.  Some of them could be addressed as follow up patches.

shefty commented:
@dshinaberry - to help with merging, any comments that are basically coding style related or non-functional changes can be handled as separate patches.  Only worry about updating the areas where comments were referencing possible issues/bugs in the code.

dshinaberry commented:
@shefty - At this point I think virtually all of the open comments are referencing possible issues/bugs. Have you completely finished re-reviewing the latest commit or are you still in progress? I think I still have unsquashed commits in my local repo. I could back out the style/formatting changes if that'd help.

shefty commented:
I finished reviewing the latest commit.  There's no need to remove any changes that you've already made.  My comment was more that you could defer making those changes, not that you needed to.  If you already have them, that's good.

shefty commented on prov/verbs/include/windows/verbs_nd.h:
```diff
@@ -0,0 +1,252 @@
+	struct nd_cm_connect_event connect_event;
+	struct nd_cm_connect_event peer_event;
+	IND2Listener *listener;
+    struct nd_cm_listen_event listen_event;
```
nit nit: indentation

shefty commented on prov/verbs/src/windows/verbs_nd_ibv.c:
```diff
@@ -0,0 +1,918 @@
+
+		ofi_mutex_lock(&cq_nd->lock);
+		while (true) {
+			if (cq_nd->notify_pending) {
```
nit:
```
while (cq_nd->notify_pending) {
    ofi_mutex_unlock(...)
    ...
}
```

shefty reviewed (approved):
I think there are some minor cleanups that could be done, but this is a great start to a more usable/maintainable provider for windows.  Merging.  Please add any cleanups as follow up patches.

shefty commented:
I should have added that we should add this provider to the windows CI builds.

dshinaberry commented:
Thanks for the merge. It's my first significant contribution to an open source project. :-)

I have some additional changes in the pipeline due to issues uncovered by implementing fabtests on windows. I'll cleanup the nits you mentioned.

I'm not sure what adding verbs to the windows CI builds requires beyond the merge. I would have thought that since the CI just builds the various configurations in the libfabric.sln file and the verbs files are in all the configurations except the efa specific ones that this provider would be built under the windows CI builds.

shefty commented:
I'll check when a new PR is opened to see if this is already building by default, or if something else is needed.
